src/components/MediaLibrary/MediaLibrary.js

```diff
@@ -0,0 +1,332 @@
+import Dialog from '../UI/Dialog';
+import { resolvePath } from '../../lib/pathHelper';
+import { changeDraftField } from '../../actions/entries';
+import {
+  loadMedia as loadMediaAction,
+  persistMedia as persistMediaAction,
+  deleteMedia as deleteMediaAction,
+  insertMedia as insertMediaAction,
+  closeMediaLibrary as closeMediaLibraryAction,
+} from '../../actions/mediaLibrary';
+import MediaLibraryFooter from './MediaLibraryFooter';
+
+/**
+ * Extensions used to determine which files to show when the media library is
+ * accessed from an image insertion field.
+ */
+const IMAGE_EXTENSIONS_VIEWABLE = [ 'jpg', 'jpeg', 'webp', 'gif', 'png', 'bmp', 'tiff' ];
+const IMAGE_EXTENSIONS = [ ...IMAGE_EXTENSIONS_VIEWABLE, 'svg' ];
+
+class MediaLibrary extends React.Component {
+
+  /**
+   * The currently selected file and query are tracked in component state as
+   * they do not impact the rest of the application.
+   */
+  state = {
+    selectedFile: {},
+    query: '',
+  };
+
+  componentDidMount() {
+    this.props.loadMedia();
+  }
+
+  componentWillReceiveProps(nextProps) {
+    /**
+     * We clear old state from the media library when it's being re-opened
+     * because, when doing so on close, the state is cleared while the media
+     * library is still fading away.
+     */
+    const isOpening = !this.props.isVisible && nextProps.isVisible;
+    if (isOpening) {
+      this.setState({ selectedFile: {}, query: '' });
+    }
+  }
+
+  /**
+   * Filter an array of file data to include only images.
+   */
+  filterImages = files => {
+    return files ? files.filter(file => IMAGE_EXTENSIONS.includes(last(file.name.split('.')))) : [];
+  };
+
+  /**
+   * Transform file data for table display.
+   */
+  toTableData = files => {
+    const tableData = files && files.map(({ key, name, size, queryOrder, url, urlIsPublicPath }) => {
+      const ext = last(name.split('.'));
+      return {
+        key,
+        name,
+        type: ext.toUpperCase(),
+        size,
+        queryOrder,
+        url,
+        urlIsPublicPath,
+        isImage: IMAGE_EXTENSIONS.includes(ext),
+        isViewableImage: IMAGE_EXTENSIONS_VIEWABLE.includes(ext),
+      };
+    });
+
+    /**
+     * Get the sort order for use with `lodash.orderBy`, and always add the
+     * `queryOrder` sort as the lowest priority sort order.
+     */
+    const { sortFields } = this.state;
+    const fieldNames = map(sortFields, 'fieldName').concat('queryOrder');
+    const directions = map(sortFields, 'direction').concat('asc');
+    return orderBy(tableData, fieldNames, directions);
+  };
+
+  handleClose = () => {
+    this.props.closeMediaLibrary();
+  };
+
+  /**
+   * Toggle asset selection on click.
+   */
+  handleAssetClick = asset => {
+    const selectedFile = this.state.selectedFile.key === asset.key ? {} : asset;
+    this.setState({ selectedFile });
+  };
+
+  /**
+   * Upload a file.
+   */
+  handlePersist = async event => {
```

Comment:
Didn't know you could use async with a naked argument in an arrow function - cool.